aptos-move/move-examples/token_objects/sources/hero.move

```diff
@@ -0,0 +1,287 @@
+module token_objects::hero {
+    use std::error;
+    use std::option::{Self, Option};
+    use std::signer;
+    use std::string::{Self, String};
+
+    use aptos_framework::object::{Self, CreatorRef, ObjectId};
+
+    use token_objects::collection;
+    use token_objects::token;
+
+    const ENOT_A_HERO: u64 = 1;
+    const ENOT_A_WEAPON: u64 = 2;
+    const ENOT_A_GEM: u64 = 3;
+    const ENOT_CREATOR: u64 = 4;
+    const EINVALID_WEAPON_UNEQUIP: u64 = 5;
+    const EINVALID_GEM_UNEQUIP: u64 = 6;
+
+    struct OnChainConfig has key {
+        collection: String,
+        mutability_config: token::MutabilityConfig,
+    }
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Hero has key {
+        armor: Option<ObjectId>,
```

Comment:
Here is where it really would be important to have typed references: `amor: Option<Ref<Armor>>`. The `Ref<T>` would guarantee that the resource actually exists. This seems to be practically an important invariant. One can easily mess it up when working with multiple objects when they are untyped (swaping them etc.)

Reply:
The reason we cannot necessarily have that is because what happens if the creator deletes the object? I guess if Ref is part of the adapter, those can be detected at load time and appropriately designated as does not exist -- or we can improve the expression.

Honestly this is where I spent most of my tension trying to make something that was both correct and expressive.

Reply:
Either that or we could introduce WeakRef for deletable objects. I think it is paramount that we allow for deletion, otherwise we could see a proliferation of burned NFTs that cannot be cleaned up and objects cannot be a fundamental unit for us. FWIW, ObjectId is intended to be a weakref for now.

Curious if you feel like this is something that's required or something that would be beneficial. I think this is purely additive, important quality of life and critical for better abstractions, but I don't see it as required.

Reply:
I agree with @wrwg 's reasoning here. If you have a set of modules that leverages objects for everything you end up passing around many untyped object ids which can easily introduce bugs that are only caught at runtime (=bad devex + bugs) and worst case you might equip a Boot as Weapon to a Hero which breaks all kind of assumptions in downstream systems like indexer. Overall this feels like a step back, both in terms of and against the core safety philosophy of Move.

I also agree we need deletion though.

Reply:
Just do add I think having `Option<ObjectId>` everywhere also severely reduces readability/understandability of complex code.